optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsAbstractBiConstraintStream.java

```diff
@@ -0,0 +1,330 @@
+            ToIntBiFunction<A, B> matchWeigher, boolean positively) {
+        DroolsConstraint<Solution_> constraint = buildConstraintConfigurable(constraintPackage, constraintName,
+                positively);
+        DroolsScoringBiConstraintStream<Solution_, A, B> stream =
+                new DroolsScoringBiConstraintStream<>(constraintFactory, this, constraint, matchWeigher);
+        childStreamList.add(stream);
+        return constraint;
+    }
+
+    protected final Constraint impactScoreConfigurableLong(String constraintPackage, String constraintName,
+            ToLongBiFunction<A, B> matchWeigher, boolean positively) {
+        DroolsConstraint<Solution_> constraint = buildConstraintConfigurable(constraintPackage, constraintName,
+                positively);
+        DroolsScoringBiConstraintStream<Solution_, A, B> stream =
+                new DroolsScoringBiConstraintStream<>(constraintFactory, this, constraint, matchWeigher);
+        childStreamList.add(stream);
+        return constraint;
+    }
+
+    protected final Constraint impactScoreConfigurableBigDecimal(String constraintPackage, String constraintName,
+            BiFunction<A, B, BigDecimal> matchWeigher, boolean positively) {
+        DroolsConstraint<Solution_> constraint = buildConstraintConfigurable(constraintPackage, constraintName,
+                positively);
+        DroolsScoringBiConstraintStream<Solution_, A, B> stream =
+                new DroolsScoringBiConstraintStream<>(constraintFactory, this, constraint, matchWeigher);
+        childStreamList.add(stream);
+        return constraint;
+    }
+
+    // ************************************************************************
+    // Pattern creation
+    // ************************************************************************
+
+    @Override
+    public void createRuleItemBuilders(List<RuleItemBuilder<?>> ruleItemBuilderList,
+            Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
+        for (DroolsAbstractBiConstraintStream<Solution_, A, B> childStream : childStreamList) {
+            childStream.createRuleItemBuilders(ruleItemBuilderList, scoreHolderGlobal);
+        }
+    }
+
+    @Override
+    public List<DroolsFromUniConstraintStream<Solution_, Object>> getFromStreamList() {
+        if (parent == null) {
+            DroolsJoinBiConstraintStream<Solution_, A, B> joinStream =
+                    (DroolsJoinBiConstraintStream<Solution_, A, B>) this;
+            List<DroolsFromUniConstraintStream<Solution_, Object>> leftParentFromStreamList =
+                    joinStream.getLeftParentStream().getFromStreamList();
+            List<DroolsFromUniConstraintStream<Solution_, Object>> rightParentFromStreamList =
+                    joinStream.getRightParentStream().getFromStreamList();
+            return Stream.concat(leftParentFromStreamList.stream(), rightParentFromStreamList.stream())
+                    .distinct()
```

Comment:
1) Should the distinct happen here? I doubt it:
I'd expect the first caller of getFromStreamList() to have to do the distinct too, so there it's a perf loss to do it in between too.

2) Is the distinct worth anything? Do the From's implement equals/hashcode? When in doubt, leave it out and put in a "TODO Should we distinct?".